infrastructure/keymanager/src/keymanager.rs

```diff
@@ -0,0 +1,262 @@
+    keys::SecretKeyFactory,
+    ristretto::RistrettoSecretKey as SecretKey,
+};
+use derive_error::Error;
+use mnemonic::*;
+use rand;
+use std::{
+    fs::File,
+    io::{prelude::*, ErrorKind},
+};
+
+#[derive(Debug, Error)]
+pub enum KeyManagerError {
+    // Could not convert into byte array
+    ByteArrayError,
+    // Could not convert provided Mnemonic into master key
+    DecodeMnemonic,
+    // The specified backup file could not be opened
+    FileOpen,
+    // Could not read from backup file
+    FileRead,
+    // Problem deserializing JSON into a new KeyManager
+    Deserialize,
+}
+
+impl From<ByteArrayError> for KeyManagerError {
+    /// Converts from ByteArrayError to KeyManagerError
+    fn from(_e: ByteArrayError) -> Self {
+        KeyManagerError::ByteArrayError
+    }
+}
+
+impl From<MnemonicError> for KeyManagerError {
+    /// Converts from MnemonicError to KeyManagerError
+    fn from(_e: MnemonicError) -> Self {
+        KeyManagerError::DecodeMnemonic
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DerivedKey {
+    pub k: SecretKey,
+    pub key_index: usize,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct KeyManager {
+    pub master_key: SecretKey,
+    pub branch_seed: String,
+    pub primary_key_index: usize,
+}
+
+impl KeyManager {
+    /// Creates a new KeyManager with a new randomly selected master_key
+    pub fn new() -> KeyManager {
+        let mut rng = rand::OsRng::new().unwrap();
+        KeyManager { master_key: SecretKey::random(&mut rng), branch_seed: "".to_string(), primary_key_index: 0 }
+    }
+
+    /// Constructs a KeyManager from known parts
+    pub fn from(master_key: SecretKey, branch_seed: String, primary_key_index: usize) -> KeyManager {
+        KeyManager { master_key, branch_seed, primary_key_index }
+    }
+
+    /// Constructs a KeyManager by generating a master_key=SHA256(seed_phrase) using a non-mnemonic seed phrase
+    pub fn from_seed_phrase(
+        seed_phrase: String,
+        branch_seed: String,
+        primary_key_index: usize,
+    ) -> Result<KeyManager, KeyManagerError>
+    {
+        match SecretKey::from_bytes(sha256(seed_phrase.into_bytes()).as_slice()) {
+            Ok(master_key) => Ok(KeyManager { master_key, branch_seed, primary_key_index }),
+            Err(e) => Err(KeyManagerError::from(e)),
+        }
+    }
+
+    /// Creates a KeyManager from the provided sequence of mnemonic words, the language of the mnemonic sequence will be
+    /// auto detected
+    pub fn from_mnemonic(
+        mnemonic_seq: &Vec<String>,
+        branch_seed: String,
+        primary_key_index: usize,
+    ) -> Result<KeyManager, KeyManagerError>
+    {
+        match SecretKey::from_mnemonic(mnemonic_seq) {
+            Ok(master_key) => Ok(KeyManager { master_key, branch_seed, primary_key_index }),
+            Err(e) => Err(KeyManagerError::from(e)),
+        }
+    }
+
+    // TODO: file should be decrypted using Salsa20 or ChaCha20
```

Comment:
Is there a `Decrypt` trait we can use here? You can implement the default (unencrypted) very easily, but then we can bring in a decryption scheme whenever, and also keep the coupling loose

Reply:
Maybe serde has mechanics for this. I can imagine that we're agnostic about how the data is stored. An implementation could even pipe FILE -> gunzip -> decrypt , and this code wouldn't need to change

Reply:
agreed, rather impl with a trait